### PR TITLE
Bug/par 4929 magento sdk normalization issues

### PR DESCRIPTION
**Pull request by adityaagashe1**
# Description

This PR will fix following 2 issues 
1. Sometimes simple offer button was rendered on cart even if offer is already in cart
2. Normalization was not balancing the qty in accordance with merchant lineitem qty. 

## Ticket

List your ticket here

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor (new code changes but everything functions the same)

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have deployed and validated my code on a sandbox
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: view/frontend/web/js/view/cart/normalize-cart.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        })
+      }
+    } catch (error) {
+      // Swallow error to avoid impacting customer checkout experience
+      /* eslint-disable-next-line no-console */
+      console.error(error)
+    }
+  }
+
+  return function (config) {
```
Comment (aarondeane-extend) [inline view/frontend/web/js/view/cart/normalize-cart.js]: `config` supposed to be an optional parameter?

**Comment by aarondeane-extend** [comment]
@adityaagashe1 seems to be working as expected in the cart... however, the mini cart outside of checkout will still allow you to add more protection plans

File: Model/ProductProtection.php [hunk 2]
```diff
@@ -408,9 +408,12 @@ private function upsert(
             }
 
             $options = $this->createOptions($product, $item, $optionValues);
-            $item->setOptions($options);
-
-            $this->addAdditionalOptions($item, $productId, $term, $leadToken, $leadQuantity);
+            if (isset($options)) {
```
Comment (jay-faust-extend) [inline Model/ProductProtection.php]: Why would `$options` not be set? Does `isset()` check for empty as well?

File: view/frontend/templates/cart/cart-normalization.phtml
Comment (jay-faust-extend) [inline view/frontend/templates/cart/cart-normalization.phtml:1]: Why do we need this?

File: view/frontend/web/js/utils/cart-utils.js [hunk 3]
```diff
@@ -9,6 +9,9 @@ define(['Magento_Customer/js/customer-data'], function (customerData) {
     return customerData.get('cart')().items ?? []
   }
 
+  const getCartData = function () {
+    return customerData.get('cart')
+  }
```
Comment (jay-faust-extend) [inline view/frontend/web/js/utils/cart-utils.js:14]: nit: the above function could call this for consistency

File: view/frontend/web/js/view/cart/normalize-cart.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+define(['Magento_Customer/js/customer-data', 'ExtendMagento'], function (
+  customerData,
+  ExtendMagento,
+) {
+  'use strict'
+
+  function normalize() {
+    try {
+      const cartItems = customerData.get('cart')().items
```
Comment (jay-faust-extend) [inline view/frontend/web/js/view/cart/normalize-cart.js]: We have that util func for this

File: view/frontend/templates/cart/cart-normalization.phtml
Comment (adityaagashe1) [reply view/frontend/templates/cart/cart-normalization.phtml:1]: This allows registering the js component on default xml.

File: view/frontend/web/js/utils/cart-utils.js (hunk 3 above)
Comment (adityaagashe1) [reply view/frontend/web/js/utils/cart-utils.js:14]: Keeping it this way. This function is used at many places.

File: view/frontend/web/js/view/cart/normalize-cart.js (hunk 4 above)
Comment (adityaagashe1) [reply view/frontend/web/js/view/cart/normalize-cart.js]: Changed to use util function.

File: view/frontend/web/js/view/cart/normalize-cart.js (hunk 1 above)
Comment (adityaagashe1) [reply view/frontend/web/js/view/cart/normalize-cart.js]: Removed config

File: view/frontend/web/js/utils/cart-utils.js (hunk 3 above)
Comment (jay-faust-extend) [reply view/frontend/web/js/utils/cart-utils.js:14]: Just to be clear I meant the above `getCartItems()` could call `getCartData()` and return its `items`.  Fine the way it is though.

**Review by jay-faust-extend — DISMISSED** [review]

File: Model/ProductProtection.php (hunk 2 above)
Comment (adityaagashe1) [reply Model/ProductProtection.php]: Removed this check

**Review by milesmcleod — DISMISSED** [review]

File: view/frontend/web/js/view/cart/normalize-cart.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      // Swallow error to avoid impacting customer checkout experience
+      /* eslint-disable-next-line no-console */
+      console.error(error)
+    }
+  }
+
+  return function () {
+    try {
+      // Normalize on cart changes
+      customerData.get('cart').subscribe(function () {
```
Comment (jay-faust-extend) [inline view/frontend/web/js/view/cart/normalize-cart.js]: Should be able to use the util func and remove the `Magento_Customer/js/customer-data` import

**Review by jay-faust-extend — APPROVED** [review]

**Comment by aarondeane-extend** [comment]
Everything appears to be working as expected now @adityaagashe1 and no more multiple calls! Great work!

**Review by aarondeane-extend — APPROVED** [review]

**Review by milesmcleod — APPROVED** [review]

**Pull request merged**